server/src/test/java/org/elasticsearch/cluster/ClusterStateIT.java

```diff
@@ -0,0 +1,351 @@
+
+        private static final ParseField VALUE = new ParseField("value");
+
+        private final int value;
+
+        int value() {
+            return value;
+        }
+
+        Custom(final int value) {
+            this.value = value;
+        }
+
+        Custom(final StreamInput in) throws IOException {
+            value = in.readInt();
+        }
+
+        @Override
+        public EnumSet<MetaData.XContentContext> context() {
+            return MetaData.ALL_CONTEXTS;
+        }
+
+        @Override
+        public Diff<MetaData.Custom> diff(final MetaData.Custom previousState) {
+            return null;
+        }
+
+        @Override
+        public void writeTo(final StreamOutput out) throws IOException {
+            out.writeInt(value);
+        }
+
+        public static int fromXContent(final XContentParser parser) throws IOException {
+            XContentParser.Token token;
+            int value = 0;
+            String currentFieldName = null;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+                switch (token) {
+                    case FIELD_NAME:
+                        currentFieldName = parser.currentName();
+                        break;
+                    case VALUE_BOOLEAN:
+                        if (VALUE.match(currentFieldName, parser.getDeprecationHandler())) {
+                            value = parser.intValue();
+                        }
+                        break;
+                }
+            }
+            return value;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(VALUE.getPreferredName(), value);
+            return builder;
+        }
+
+    }
+
+    public static class NodeCustom extends Custom {
+
+        public static final String TYPE = "node";
+
+        NodeCustom(final int value) {
+            super(value);
+        }
+
+        NodeCustom(final StreamInput in) throws IOException {
+            super(in);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return TYPE;
+        }
+
+        @Override
+        public Optional<String> getRequiredFeature() {
+            return Optional.of("node");
+        }
+
+    }
+
+    public static class NodeAndTransportClientCustom extends Custom {
+
+        public static final String TYPE = "node-and-transport-client";
+
+        NodeAndTransportClientCustom(final int value) {
+            super(value);
+        }
+
+        public NodeAndTransportClientCustom(final StreamInput in) throws IOException {
+            super(in);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return TYPE;
+        }
+
+        @Override
+        public Optional<String> getRequiredFeature() {
+            return Optional.of("node-and-transport-client");
```

Comment:
shall we sometime return non existing value here? (or have yet another type)

Reply:
I am not sure what you're going for here; the required feature do not have to line up with the custom type name, they just happen to do that in this test but they are orthogonal. Would it help if I used different names for the required feature and the custom type to make this clear?

Reply:
I don't think we currently test a custom that returns no required features in this test. Since this custom is expected to be returned all the time which is the same behavior as having no required features, I wonder if we sometime want to return an empty Optional. My bad for call it "non existing" - I can see how that may be read differently.

Reply:
I pushed 546fe414e5af5ad8858d54a7a202c6196524fa36.